src/backend/replication/walproposer.c

```diff
@@ -1233,13 +1230,175 @@ WalProposerPoll(void)
+	WalMessage *msg;
+	AppendRequestHeader *req;
+
+	if (wk->flushWrite)
+	{
+		if (!AsyncFlush(wk))
+			/* 
+			 * AsyncFlush failed, that could happen if the socket is closed or
+			 * we have nothing to write and should wait for writeable socket.
+			 */
+			return wk->state == SS_ACTIVE;
+
+		wk->currMsg = wk->currMsg->next;
+		wk->flushWrite = false;
+	}
+
+	while (wk->currMsg)
+	{
+		msg = wk->currMsg;
+		req = &msg->req;
+
+		req->commitLsn = GetAcknowledgedByQuorumWALPosition();
+		req->truncateLsn = truncateLsn;
+
+		Assert((msg->ackMask & (1 << wki)) == 0);
+
+		/*
+		 * If we need to send this message not from the beginning,
+		 * form the cut version. Only happens for the first
+		 * message.
+		 */
+		if (wk->startStreamingAt > msg->req.beginLsn)
+		{
+			uint32		len;
+			uint32		size;
+
+			Assert(wk->startStreamingAt < req->endLsn);
+
+			len = msg->req.endLsn - wk->startStreamingAt;
+			size = sizeof(AppendRequestHeader) + len;
+			req = malloc(size);
+			*req = msg->req;
+			req->beginLsn = wk->startStreamingAt;
+			memcpy(req + 1,
+					(char *) (&msg->req + 1) + wk->startStreamingAt -
+					msg->req.beginLsn,
+					len);
+		}
+
+		elog(LOG,
+				"sending message len %ld beginLsn=%X/%X endLsn=%X/%X commitLsn=%X/%X truncateLsn=%X/%X to %s:%s",
+				req->endLsn - req->beginLsn,
+				LSN_FORMAT_ARGS(req->beginLsn),
+				LSN_FORMAT_ARGS(req->endLsn),
+				LSN_FORMAT_ARGS(req->commitLsn),
+				LSN_FORMAT_ARGS(truncateLsn), wk->host, wk->port);
+
+		/* if this is the first sent message, we should start processing feedback */
+		if (wk->ackMsg == NULL)
+			wk->ackMsg = wk->currMsg;
+
+		/*
+		 * We write with msg->size here because the body of the
+		 * message is stored after the end of the WalMessage
+		 * struct, in the allocation for each msg
+		 */
+		if (!AsyncWrite(wk, req,
+						sizeof(AppendRequestHeader) + req->endLsn - req->beginLsn,
+						SS_ACTIVE))
+		{
+			if (req != &msg->req)
+				free(req);
+			if (wk->state == SS_ACTIVE)
+			{
+				wk->flushWrite = true;
+				return true;
+			}
+			return false;
+		}
+		if (req != &msg->req)
+			free(req);
+
+		/* continue writing the next message */
+		wk->currMsg = wk->currMsg->next;
+	}
+
+	return true;
+}
+
+/*
+ * Receive and process all available feedback.
+ *
+ * Can change state if Async* functions encounter errors and reset connection.
+ * Returns false in this case, true otherwise.
+ * 
+ * NB: This function can call SendMessageToNode and produce new messages.
+ */
+static bool
+RecvAppendResponses(WalKeeper *wk)
+{
+	XLogRecPtr	minQuorumLsn;
+	int wki = wk - walkeeper;
+	bool readAnything = false;
+
+	/*
+	 * We shouldn't read responses ahead of wk->currMsg, because that will
+	 * look like we are receiving responses for messages that haven't been
+	 * sent yet. This can happen when message was placed in a buffer in 
+	 * SendAppendRequests, but sent through a wire only with a flush in 
+	 * reading routine.
+	 */
+	while (wk->ackMsg != NULL && wk->ackMsg != wk->currMsg)
```

Comment:
With this coding, if we don't wait for anything and connection is lost, we won't try reading from it and thus will never reset it, right? So better prepare to async responses right away and just expect here any ack message.

Also this 'sent with a flush in reading routine' shouldn't happen anymore...

Reply:
'sent with a flush in reading routine' still can happen, it's in PQConsumeInput implementation: https://github.com/zenithdb/postgres/blob/main/src/interfaces/libpq/fe-exec.c#L1909-L1919

As we handle all read/write events and don't need to flush in PQConsumeInput, we may try to switch to using pqReadData instead. But for now we still can read responses ahead of requests, so I'd better create an issue for better async handling and fix it later.